File: backend/app/schemas/precheck.py
from datetime import date, time

from pydantic import BaseModel, Field


class PreCheckRequest(BaseModel):
    department_id: str
    section_id: str

    location_id: str | None = None
    asset_id: str | None = None
    maintenance_task_id: str | None = None

    preferred_date: date
    preferred_start_time: time
    preferred_end_time: time

    minimum_duration_minutes: int = Field(
        gt=0,
        le=1440,
    )

    block_type: str = Field(
        min_length=1,
        max_length=50,
    )

    power_block_required: bool = False

    resource_ids: list[str] = []


class PreCheckIssue(BaseModel):
    code: str
    severity: str
    message: str


class SuggestedWindow(BaseModel):
    start_time: time
    end_time: time
    score: float
    reason: str


class PreCheckResponse(BaseModel):
    status: str
    can_submit: bool

    checks_total: int
    checks_passed: int

    warnings: int
    blocks: int

    issues: list[PreCheckIssue]
    suggested_windows: list[SuggestedWindow]

File: frontend/src/api/precheck.ts
export type PreCheckIssue = {
  code: string
  severity: string
  message: string
}

export type SuggestedWindow = {
  start_time: string
  end_time: string
  score: number
  reason: string
}

export type PreCheckRequest = {
  department_id: string
  section_id: string
  location_id?: string
  asset_id?: string
  maintenance_task_id?: string
  preferred_date: string
  preferred_start_time: string
  block_request_id?: string
  preferred_end_time: string
  minimum_duration_minutes: number
  block_type: string
  power_block_required: boolean
  resource_ids: string[]
}

export type PreCheckResponse = {
  status: string
  can_submit: boolean
  checks_total: number
  checks_passed: number
  warnings: number
  blocks: number
  issues: PreCheckIssue[]
  suggested_windows: SuggestedWindow[]
}

const API = 'http://127.0.0.1:8000'

export async function runPreCheck(
  request: PreCheckRequest,
): Promise<PreCheckResponse> {
  const response = await fetch(
    `${API}/api/v1/precheck`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    },
  )

  if (!response.ok) {
    throw new Error(
      `Pre-check failed: ${response.status}`,
    )
  }

  return response.json()
}